src/transforms/sample/transform.rs

```diff
@@ -12,24 +12,112 @@ use crate::{
 use vector_lib::lookup::lookup_v2::OptionalValuePath;
 use vector_lib::lookup::OwnedTargetPath;
 
+/// Exists only for backwards compatability purposes so that the value of sample_rate_key is
+/// consistent after the internal implementation of the Sample class was modified to work in terms
+/// of percentages
+#[derive(Clone, Debug)]
+pub enum SampleMode {
+    Rate {
+        rate: u64,
+        counters: HashMap<Option<String>, u64>,
+    },
+    Ratio {
+        ratio: f64,
+        values: HashMap<Option<String>, f64>,
+        hash_ratio_threshold: u64,
+    },
+}
+
+impl SampleMode {
+    pub fn new_rate(rate: u64) -> Self {
+        Self::Rate {
+            rate,
+            counters: HashMap::default(),
+        }
+    }
+
+    pub fn new_ratio(ratio: f64) -> Self {
+        Self::Ratio {
+            ratio,
+            values: HashMap::default(),
+            // Supports the 'key_field' option, assuming an equal distribution of values for a given
+            // field, hashing its contents this component should output events according to the
+            // configured ratio.
+            //
+            // To do this convert the hash to a number between 0 and 1 and compare to the ratio, to
+            // address issues with precision here the ratio is expanded to meet the width of the type of
+            // the hash.
+            hash_ratio_threshold: (ratio * ((u64::MAX as u128) + 1) as f64) as u64,
+        }
+    }
+
+    fn increment(&mut self, group_by_key: &Option<String>, value: &Option<Cow<'_, str>>) -> bool {
+        let threshold_exceeded = match self {
+            Self::Rate { rate, counters } => {
+                let counter_value = counters.entry(group_by_key.clone()).or_default();
+                let old_counter_value = *counter_value;
+                let increment: u64 = *counter_value + 1;
+                *counter_value = increment;
+                old_counter_value % *rate == 0
+            }
+            Self::Ratio { ratio, values, .. } => {
+                let value = values.entry(group_by_key.clone()).or_insert(1.0 - *ratio);
+                let increment: f64 = *value + *ratio;
+                *value = if increment >= 1.0 {
+                    increment - 1.0
+                } else {
+                    increment
+                };
+                increment >= 1.0
+            }
+        };
+        if let Some(value) = value {
+            self.hash_within_ratio(value.as_bytes())
+        } else {
+            threshold_exceeded
+        }
+    }
+
+    fn hash_within_ratio(&self, value: &[u8]) -> bool {
+        let hash = seahash::hash(value);
+        match self {
+            Self::Rate { rate, .. } => hash % rate == 0,
+            Self::Ratio {
+                hash_ratio_threshold,
```

Comment:
formatting nit: I think if the word here has only two parts instead of three (ex `hash_threshold`), `rustfmt` won't split the line up.